Client/qt/Dialog.cpp

```diff
@@ -0,0 +1,134 @@
+	connectButton->setEnabled(false);
+	deviceDiscoveryGroupBox->setEnabled(false);
+	ambientSoundModeGroupBox->setEnabled(true);
+	ambientSoundControlCheckBox->setEnabled(true);
+	// Send initial state
+	sendCommandFuture.setFromAsync([this]() {
+		return btWrap->sendCommand(CommandSerializer::serializeNcAndAsmSetting(
+			NC_ASM_EFFECT::ADJUSTMENT_COMPLETION,
+			NC_ASM_SETTING_TYPE::LEVEL_ADJUSTMENT,
+			ASM_SETTING_TYPE::LEVEL_ADJUSTMENT,
+			ASM_ID::NORMAL,
+			0));
+	});
+	timer->start(100);
+}
+
+void Dialog::on_timeout() {
+	bool ambientSoundControl = ambientSoundControlCheckBox->isChecked();
+	static bool sentAmbientSoundControl = ambientSoundControl;
+	bool focusOnVoice = focusOnVoiceCheckBox->isChecked();
+	static bool sentFocusOnVoice = focusOnVoice;
+	int asmLevel = ambientSoundSlider->value();
+	static int lastAsmLevel = asmLevel;
+	static int sentAsmLevel = asmLevel;
+#ifndef NDEBUG
+	qDebug() << "ambientSoundControl:" << sentAmbientSoundControl << "<>"
+			 << ambientSoundControl;
+	qDebug() << "focusOnVoice:" << sentFocusOnVoice << "<>" << focusOnVoice;
+	qDebug() << "asmLevel:" << sentAsmLevel << "<>" << asmLevel;
+	qDebug() << "lastAsmLevel =" << lastAsmLevel;
+#endif
+	if (sendCommandFuture.ready()) {
+		try {
+			sendCommandFuture.get();
+		} catch (const RecoverableException &exc) {
+			if (exc.shouldDisconnect) {
+				btWrap->disconnect();
+				qDebug() << "Disconnected due to" << exc.what();
+			}
+		}
+	} else if (sendCommandFuture.valid()) {
+		qDebug() << "Waiting for prior command to send";
+	} else if (sentAsmLevel != asmLevel || sentFocusOnVoice != focusOnVoice ||
+			   sentAmbientSoundControl != ambientSoundControl) {
+		qDebug() << "Sending command";
+		auto ncAsmEffect = ambientSoundSlider->isSliderDown() ?
+								 NC_ASM_EFFECT::ADJUSTMENT_IN_PROGRESS :
+						   ambientSoundControl ?
+								 NC_ASM_EFFECT::ADJUSTMENT_COMPLETION :
+								 NC_ASM_EFFECT::OFF;
+		auto asmId = focusOnVoice ? ASM_ID::VOICE : ASM_ID::NORMAL;
+		lastAsmLevel = asmLevel == -1 ? lastAsmLevel : asmLevel;
+		asmLevel = ambientSoundControl ? lastAsmLevel : -1;
+		sendCommandFuture.setFromAsync([this, asmId, ncAsmEffect, asmLevel]() {
+			return btWrap->sendCommand(
+				CommandSerializer::serializeNcAndAsmSetting(
+					ncAsmEffect,
+					NC_ASM_SETTING_TYPE::LEVEL_ADJUSTMENT,
+					ASM_SETTING_TYPE::LEVEL_ADJUSTMENT,
+					asmId,
+					asmLevel));
+		});
+		sentAsmLevel = asmLevel;
+		sentFocusOnVoice = focusOnVoice;
+		sentAmbientSoundControl = ambientSoundControl;
+	}
+}
+
+void Dialog::on_ambientSoundSlider_valueChanged(int value) {
+	focusOnVoiceCheckBox->setEnabled(value >= MINIMUM_VOICE_FOCUS_STEP);
+}
+
+void Dialog::on_deviceListWidget_itemSelectionChanged() {
+	if (!deviceListWidget->selectedItems().isEmpty()) {
+		selectedDevice = deviceListWidget->selectedItems().first()->text();
+		connectButton->setEnabled(true);
+	}
+}
+
+void Dialog::on_buttonBox_clicked(QAbstractButton *_) {
+	close();
```

Comment:
Why is a separate close button needed? We don't really need to perform any cleanup (well, we can disconnect gracefully but it hardly matters) and it just means there are two close buttons :(

Reply:
Yes, I can remove the bottom close button.